### ISSUE TITLE
Clarrification on adding payloads and adjusting the thrust scaler

### ISSUE THREAD

**Issue by hildebrandt-carl** [unl-nimbus-lab/Freyja#15]
Hi,

I went through the steps for calibrating `thrust_scaler`, documented [here](https://github.com/unl-nimbus-lab/Freyja/wiki/Thrust-Calibration). The mass of the drone and battery was 1.2kg, and thus I set `total_mass=1.2`. Starting with `thrust_scaler=200` I iteratively updated it until it reached the correct set-point of `z=1`, with a final value of `thrust_scaler=102`.

I then added weight to the drone. I added exactly 500g to the drone, and thus changed the `total_mass=1.7kg`. However I noticed that running the same test, the drone only reaches `z=0.5`. 

My instinct tells me to rerun the iterative process of finding a new `thrust_scaler`. However reading more into the documentation it mentions this:
> This does not include adding payloads (you should adjust total_mass parameter for that.

So I now a bit lost as the best move forward. More generally I notice that there are several `params` in the launch file:
```xml
  <node name="lqg_controller" pkg="lqg_control" type="lqg_control_node" output="screen" if="$(eval not use_velctrl_only)">
	<param name="controller_rate"     type="int"    value="45"  />
  	<param name="total_mass"          type="double" value="$(arg total_mass)"/>
  	<param name="use_stricter_gains"  type="bool"   value="false" />
  	<param name="estimator_rate"      type="int"    value="50" />
  	<param name="bias_compensation"   type="string" value="$(arg bias_compensation)" />
  	<param name="mass_estimation"     type="bool"   value="true" />
  	<param name="mass_correction"     type="bool"   value="false" />
  	<param name="enable_flatness_ff"  type="bool"   value="$(arg enable_flatness_ff)" />
  </node>
  ```
  
  For example `mass_correction`, `bias_compensation`, `total_mass`. Is there documentation anywhere as to what each do? Also any suggestions about my current issue with changing the weight?

**Comment by ajshank** [comment]
Thanks for reporting this, @hildebrandt-carl.
For your second point: `total_mass` and `bias_compensation` should be self-explanatory, although the documentation is poor on the bias compensation part. The latter basically estimates [usually time-varying] external biases and adjusts the controller to tackle them. You could use this for compensating for the added weight, but it _should_ really already be handled by `total_mass`. The other parameter, `mass_correction`, will estimate the actual mass in the air (assuming that your `thrust_scaler` is correct) and override the provided `total_mass` parameter. This can be tricky and contentious, depending on the kind of build/frame you have. Note that `mass_estimation` is `true`, so you could always check the bags to see what the controller "thinks" the mass is.

On to your first part.
This is slightly complicated, but essentially `thrust_scaler` assumes a linear thrust response from the vehicle (if commanding "20%" was enough to produce 10N, then we must command "40%" to produce 20N), which is obviously not true -- but most autopilots accept only this % value as input, not SI units. However, ArduCopter does _some_ thrust linearisation, and this can be affected by several parameters.

Your process (re: determine `thrust_scaler` only once, and edit `total_mass` subsequently if mass changes) is correct, though. Ideally, changing the `total_mass` is all you'd need to do. So, I'm curious to look into this. Could you report the following -
1. Did you use a fresh battery after adding the weight? What is the battery voltage like?
2. (Assuming you are using ArduCopter) Have you set `MOT_BAT*` parameters to min/max ranges?
3. With `bias_compensation` set to "off" in the experiment you've described above, what is the estimated mass (from the topic `/freyja_estimated_mass`)? Is it close to 1.7 as you expect?

**Comment by ajshank** [comment]
Also, hope this new page helps: https://github.com/unl-nimbus-lab/Freyja/wiki/Bias-Compensation.

**Comment by hildebrandt-carl** [comment]
Sorry for the late reply. This is still on my queue. I will get back to you in the next few days.

**Comment by hildebrandt-carl** [comment]
Hi @ajshank ,

Thanks for the quick reply and sorry for this wall of text. I ran further experiments and collected more details while trying to answer your questions.

I had two drones for my experiments, both with _freshly charged batteries_. Both drones carry payloads with similar weight distributions. The details for both drones are as follows:

**Drone 1**: 
```
Name: JOZI
Weight: 1855g
Battery Voltage: 16.86V
MOT_BAT_CURR_MAX: 0.000A
MOT_BAT_CURR_TC: 5.000s
MOT_BAT_IDX: First battery
MOT_BAT_VOLT_MAX: 0.000V
MOT_BAT_VOLT_MIN: 0.000V
```

**Drone 2**: 
```
Name: STARK
Weight: 1833g
Battery Voltage: 16.85V
MOT_BAT_CURR_MAX: 0.000A
MOT_BAT_CURR_TC: 5.000s
MOT_BAT_IDX: First battery
MOT_BAT_VOLT_MAX: 0.000V
MOT_BAT_VOLT_MIN: 0.000V
```

I ran 3 experiments with each drone. Each experiment requested the drone to hover at the **origin**, with `z=1m`. Each experiment differed in that I changed the `bias_compensation` between `auto`, `on`, and `off`. Thus my launch file looked like this (where `drone_name` and `bias_setting` varies based on the experiment):

```xml
  <include file="$(find flight_controller)/launch/freyja_controller.launch">
    <arg name="use_waypoint_handler"  value="true"/>
    <arg name="vicon_topic"           value="/vicon/{drone_name}/{drone_name}"/>
    <arg name="total_mass"            value="1.85"/>
    <arg name="thrust_scaler"         value="70"/>
    <arg name="bias_compensation"     value="{bias_setting}"/>
  </include>
```

The results of my experiments are as follows:

----------------------------
_**Experiment 1**_:
Configuration:
`drone_name`: `JOZI`
`bias_settings`: `auto`

Results:
Small oscillation at the begging then hovers perfectly. 

Video:
https://user-images.githubusercontent.com/26097148/153071093-2d053ce2-aec8-4f63-b890-87fd20c1e1c3.mp4

Bag file:
[JOZI_auto.bag.zip](https://github.com/unl-nimbus-lab/Freyja/files/8027284/JOZI_auto.bag.zip)

----------------------------

_**Experiment 2**_:
Configuration:
`drone_name`: `JOZI`
`bias_settings`: `on`

Results:
Hovers perfectly. 

Video:
https://user-images.githubusercontent.com/26097148/153071162-04fc2af6-5da4-44f1-b2b4-fcb3705b1644.mp4

Bag file:
[JOZI_on.bag.zip](https://github.com/unl-nimbus-lab/Freyja/files/8027287/JOZI_on.bag.zip)

----------------------------

_**Experiment 3**_:
Configuration:
`drone_name`: `JOZI`
`bias_settings`: `off`

Results:
Hovers perfectly, however, does not reach the target of `z=1`. **Note** the shoot-up near the end is me switching back to manual control.

Video:
https://user-images.githubusercontent.com/26097148/153071236-1db33555-0f60-402e-a3b9-1a76e3360cd9.mp4

Bag file:
[JOZI_off.bag.zip](https://github.com/unl-nimbus-lab/Freyja/files/8027291/JOZI_off.bag.zip)

----------------------------

_**Experiment 4**_:
Configuration:
`drone_name`: `STARK`
`bias_settings`: `auto`

Results:
Large oscillations in the x and y plane. 

Video:
https://user-images.githubusercontent.com/26097148/153071305-c8054f99-f95f-44ad-b255-16e77caf6e1a.mp4

Bag file:
[STARK_auto.bag.zip](https://github.com/unl-nimbus-lab/Freyja/files/8027294/STARK_auto.bag.zip)

----------------------------

_**Experiment 5**_:
Configuration:
`drone_name`: `STARK`
`bias_settings`: `on`

Results:
Large oscillations in the x and y plane. 

Video:
https://user-images.githubusercontent.com/26097148/153071365-68e1829c-7730-449b-8093-6e2d60b92072.mp4

Bag file:
[STARK_on.bag.zip](https://github.com/unl-nimbus-lab/Freyja/files/8027299/STARK_on.bag.zip)

----------------------------

_**Experiment 6**_:
Configuration:
`drone_name`: `STARK`
`bias_settings`: `off`

Results:
Hovers perfectly, however, does not reach the target of `z=1`. 

Video:
https://user-images.githubusercontent.com/26097148/153071440-2b0c9d35-a4c4-4c31-963b-3f67e48b4289.mp4

Bag file:
[STARK_off_90.bag.zip](https://github.com/unl-nimbus-lab/Freyja/files/8027303/STARK_off_90.bag.zip)

----------------------------

At this point, I noticed that `STARK` did not seem to have any oscillation when the `bias_compensation` was set to `off`.  Thus to correct the `z` offset, I updated the `thrust_scalar` from `90` to `70`. The results are as follows:

----------------------------

_**Experiment 7**_:
Configuration:
`drone_name`: `STARK`
`bias_settings`: `off`
`thurst_scalar`: `70`

Results:
Hovers perfectly. 

Video:
https://user-images.githubusercontent.com/26097148/153071494-efa5c1ee-fe50-48ff-b554-38fdb34d2322.mp4

Bag file:
[STARK_off_70.bag.zip](https://github.com/unl-nimbus-lab/Freyja/files/8027305/STARK_off_70.bag.zip)

----------------------------

The interesting findings of my experiment are:
1) For some reason, **JOZI** flies perfectly under all `bias_compensation` values when in computer control mode. (Other than not reaching `z=1` when bias is off, which makes sense)
2) **STARK** flies perfectly when `bias_compensation` is set to `off` and the `thrust_scaler` is increased slightly.

My question moving forward are:
1) What is the downside of setting `bias_compensation` to `off`?

**Comment by hildebrandt-carl** [comment]
First, I want to thank @ajshank for all the help and consideration he put into solving this offline.

It turns out the results from above were not replicable, and it seemed like the drone would fly as expected for periods and then start oscillating and not recover from these oscillations.

The ultimate issue was that the communication rate between the radios was too low. In my case, I needed to follow [this guide](https://ardupilot.org/copter/docs/common-configuring-a-telemetry-radio-using-mission-planner.html) to update the settings.

Needless to say that there are probably several configurations of the radio that work flawlessly. However, here is the configuration I ended up using which fixed the issue:

**Local Radio (radio connected to the PC)**
![new_usb](https://user-images.githubusercontent.com/26097148/154690147-1aa5c982-083c-4cea-902f-ebc8cdc9a1a7.png)


**Remote Radio (radio connected to the drone)**
![new_drone](https://user-images.githubusercontent.com/26097148/154690121-f7fb6384-af79-452c-acef-df650e35c408.png)

Note: The major changes to the settings all allow a higher communication rate. Specifically I was looking at `Air Speed`, `Duty Cycle` and `Max Window (ms)`.
